aten/src/ATen/native/TensorProperties.cpp

```diff
@@ -13,6 +14,21 @@ bool is_same_size(const Tensor& self, const Tensor& other) {
   return self.sizes().equals(other.sizes());
 }
 
+bool nested_is_same_size(const Tensor& self, const Tensor& other) {
```

Comment:
This and its registration in native functions should also belong in the second PR because it is intended for nested support. If we add it here, it won't be tested until we reach the second PR anyway.

Reply:
Oh wait, I do see that you have tests for this in this PR, but still maybe better to have it in the second Pr to have everything in one place?

Reply:
It is intended for nested support but in theory someone could pass in nested grad to backward today and this would just be providing a better error message.  I can go remove the changes for this but it kinda feels like I am just rolling the ball further down the hill and make that future PR heavier and heavier

Reply:
Oh I see, maybe a third PR in between just to make sure the refactor is completely on its own (so CI is able to test that individually)?

Reply:
Am I removing all the changes in __init__.py  to a third PR as well?

Reply:
Created this stack: https://github.com/pytorch/pytorch/pull/79296

Reply:
Yeah I'd probably say as a rule of thumb just avoid cluttering the pure refactor PR, and throw everything into the 3rd PR. Optionally, if we really wanted to put ALL the refactors in one place to reduce clutter in the 3rd PR even more, maybe keep in the refactor PR the part where we refactor `grad` to use is_same_size and then create the new function `_calculate shape`, and then the only thing we would need to add for nested tenor support would be the extra branch in `_calculate_shape`, which would be in the 3rd PR.